File: mpi/solutions/mpi_bcast/mpi_bcast.f90
PROGRAM mpi_bcast_example
   USE mpi_f08
   IMPLICIT NONE

   INTEGER :: rank, value, ierr
   INTEGER :: input_unit, io_status

   CALL MPI_Init(ierr)

   CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr) ! what rank is the current processor

   IF (rank == 0) THEN
      ! if root process we read the value to broadcast from file
      OPEN(UNIT=10, FILE='input.txt', STATUS='OLD', ACTION='READ', IOSTAT=io_status)
      IF (io_status /= 0) THEN
         WRITE(*,*) 'Warning: Could not open input.txt, using default value'
         value = 42  ! Default value
      ELSE
         WRITE(*,*) 'Enter a number to broadcast:'
         READ(10, *, IOSTAT=io_status) value
         IF (io_status /= 0) THEN
            WRITE(*,*) 'Failed to read from input.txt, using default value'
            value = 42  ! Default value
         ELSE
            WRITE(*,'(A,I0)') 'Read value: ', value
         END IF
         CLOSE(10)
      END IF
   ELSE
      WRITE(*,'(A,I0,A,I0)') 'process ', rank, ': Before MPI_Bcast, value is ', value
   END IF

   ! each processor calls MPI_Bcast, data is broadcast from root processor and ends up in everyone value variable
   ! root process uses MPI_Bcast to broadcast the value, each other process uses MPI_Bcast to receive the broadcast value
   CALL MPI_Bcast(value, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, ierr)

   WRITE(*,'(A,I0,A,I0)') 'process ', rank, ': After MPI_Bcast, value is ', value

   CALL MPI_Finalize(ierr)

END PROGRAM mpi_bcast_example

File: mpi/solutions/mpi_bcast/mpi_bcast.c
#include <stdio.h>
#include "mpi.h"


////////////
//MPI_Bcast
////////////
//
// int MPI_Bcast( void *buffer, int count, MPI_Datatype datatype, int root, MPI_Comm comm )
//
// This example simply uses MPI_Bcast to broadcast a read in value to all other processes from root process
//
// example usage:
//		compile: mpicc -o mpi_bcast mpi_bcast.c
//		run: mpirun -n 4 mpi_bcast
//
int main(argc, argv)
int argc;
char **argv;
{
    int rank, value;
    FILE *input_file = NULL;

    MPI_Init(&argc, &argv);

    MPI_Comm_rank(MPI_COMM_WORLD, &rank); //what rank is the current processor

    if (rank == 0) {
        // if root process we read the value to broadcast from file
        input_file = fopen("input.txt", "r");
        if (!input_file) {
            printf("Warning: Could not open input.txt, using default value\n");
            value = 42; // Default value
        } else {
            printf("Enter a number to broadcast:\n");
            if (fscanf(input_file, "%d", &value) == 1) {
                printf("Read value: %d\n", value);
            } else {
                printf("Failed to read from input.txt, using default value\n");
                value = 42; // Default value
            }
            fclose(input_file);
        }
    } else {
        printf("process %d: Before MPI_Bcast, value is %d\n", rank, value); 
    }

    // each processor calls MPI_Bcast, data is broadcast from root processor and ends up in everyone value variable
    // root process uses MPI_Bcast to broadcast the value, each other process uses MPI_Bcast to receive the broadcast value
    MPI_Bcast(&value, 1, MPI_INT, 0, MPI_COMM_WORLD);

    printf("process %d: After MPI_Bcast, value is %d\n", rank, value);

    MPI_Finalize();

    return 0;
}